### DISCUSSION TITLE
How can I disable Harfbuzz WASM in Chrome, Firefox and other systems?

### DISCUSSION THREAD

**Discussion by Lewiscowles1986** [harfbuzz/harfbuzz#4767] [Q&A] [answered]
Hi, Someone made https://fuglede.github.io/llama.ttf/ and it's been highlighted to my attention that your font library, ships supporting arbitrary code injection and execution on users machines.

I Was previously unaware of this, and would like to disable this; Ideally via an environment variable, and know if there are ways I can test the facility has been disabled; or if to your knowledge it ships disabled.

This is not a personal or professional slight; I think it's "cool" and thank you for sharing openly; but I just really don't want that feature if possible, and would like to experiment seeing how it effects my use of computers having the feature off.

**Comment by khaledhosny** [comment, marked as answer]
WASM shaper is not enabled by default.

**Reply by Lewiscowles1986** [reply]
Does this mean that Chrome and Firefox do not enable for use?

**Reply by behdad** [reply]
No platform enables it.

**Comment by dwt** [comment]
Since this shaping feature offers the ability to for example craft contracts that show different terms before and after they are signed - could you provide some links to the rationale for this feature, or rather how you planned to ever get this safe?

So far it seems that this whole shaping feature is impossible to secure?

**Reply by behdad** [reply]
As already mentioned, this feature is not enabled in HarfBuzz, and it won't be, for any foreseeable future. It's a proof-of-concept technology preview.

What you mentioned, rendering things differently, is already possible with regular OpenType fonts.

**Reply by dwt** [reply]
@behdad Thanks for your answer, and it is good to know that this feature is just a technology demo.

I would like to argue though, that just because a security problem can also be caused by another means, is a really bad justification to bring on another way to achieve that same security problem.

**Reply by khaledhosny** [reply]
There is no “security” problem here. Controlling how text looks like is pretty much the definition of what a (computer) font is.

**Reply by dwt** [reply]
@khaledhosny Sure, how a font looks is the definition what a a font does. But as soon as that mechanism is so flexible that it can be used to display completely different sentences when the font is switched, and the algorithm is Turing-Complete,  that does become a problem. 

I am arguing, that the goal of a font rendering engine should not be to allow arbitrary changes of text, but to very specifically disallow them, so that the human is not confused about what is actually written - and that that is a very important security goal of font rendering engines. And incidentally what makes it actually really hard to write them and why we want a competent team that focusses on that problem.

If you disagree, I'm very glad that you guys specify so very clearly, because that is very helpful to know.

**Reply by behdad** [reply]
As I said, OpenType fonts are already that (and Turing-complete too). What I'm saying is not "whataboutism". If your system depends on fonts showing what the text was, your system is broken (for relying on unreliable font sources). It's an intractable problem to know when a font is being "deceptive".

**Reply by dwt** [reply]
I am hearing you saying that font rendering - while retaining the users ability to actually know what is written - is too hard to implement and therefore you excluded it from the scope of this project? I.e. an explicit non goal?

I would gladly provide a pull request to the readme that clarifies this there.

Perhaps something like this?

HarfBuzz is a text shaping engine. It primarily supports [OpenType](https://docs.microsoft.com/en-us/typography/opentype/spec/), but also [Apple Advanced Typography](https://developer.apple.com/fonts/TrueType-Reference-Manual/RM06/Chap6AATIntro.html). HarfBuzz is used in Android, Chrome, ChromeOS, Firefox, GNOME, GTK+, KDE, Qt, LibreOffice, OpenJDK, XeTeX, PlayStation, Microsoft Edge, Adobe Photoshop, Illustrator, InDesign, Godot Engine, and other places.

Despite its extensive usage and support for various font technologies, it is important to note that HarfBuzz, like other shaping engines, does not guarantee that the rendered text will accurately reflect the original ASCII or UTF-8 data points. Therefore, the security goal of a user reliably reading what the ASCII/UTF-8 data points represent is not a goal of the shaping engine.

**Reply by dwt** [reply]
To clarify, it seems very clear to you (the project) members that this is the case, but looking at the project from the outside, I would never have guessed this to be the case (or even allowed). Therefor it seems important to me to clarify.

**Reply by alerque** [reply]
@dwt This issue here is not that you don't understand this project, it's that you don't understand what a font is or does. The concern you have about "accurately reflecting the original data points" is already not a realistic expectation the way you think it is. The reason fonts need shaping at all is because they are (and have been for many decades) a set of transformation rules to **render** a given input. The same logic that allows the font to pick a ligature for the case when an `i` follows an `f` and render it with the dot in a way that doesn't conflict with the f could just as well be used to change the visual rendering of the input `will sign` to `was signed`. 

All fonts are a set of choices about how to render input data points into shapes. All fonts –even without WASM making it easier that having massive substitution tables can make the choice to arbitrarily make any shape they want out of any input. There are a plethora af fonts that substitute various input content for various outputs such as clipart or declarations or mockup blobs. There are fonts that output game UIs and update the game state based on the input text (not using WASM, just regular open type features). There are fonts that substitute specific words into other languages or renderings. There are fonts that censor certain words like profanity.

Besides some languages relying on this to even be readable, the entire ecosystem of emoji basically relies on fonts that render specific input sequences differently based on context. Flag emojis are made of of two character input sequences in a specific range, such as "U-S" that displays a US flag (🇺🇸) but "R-U" displays a Russian flag (🇷🇺). The U in the input sequence is the same both places, but the sequence of input letters produces a completely different output rendering. A malicious font could substitute any country flag (or spelled out name of a country) with any rendered shapes of its choosing. Many other emojis use skin tone modifiers or other contextual sequences to trigger certain outputs.

Your choice of font is just as important as your choice of shaper and browser and operating system. It can/will affect how you see everything. If you must know what UTF-8 data points were used (use a hex editor—and of course a font you trust to view that output because the same issue applies).

**Reply by alerque** [reply]
Maybe also it is helpful to say this: the job of the shaper is to follow the rules of the font, not enforce some arbitrary ethics of it's own. A font that takes the input U+0041 and outputs a glyph shaped like a B instead of an A is somewhat objectively buggy. The bug is in the font, not in the shaper. The shaper doesn't validate that the output has something roughly resembling an English capital A. It might just as well be a round blob that the font used as a placeholder for UI mockup purposes.

Likewise a font that mistakenly returns the wrong diacritic mark when presented with a base letter and a combining diacritic in an input is a bug in the font not the shaper.

Likewise it is not an attack vector the shaper can reasonable be expected to "protect" against. If the font says that an "A" should look like "this series of vector shapes" and those shapes look like a "B", that is the font's problem.

If the font decides that the input sequence U+1F1F0 followed by U+1F1FF should be rendered with a green tinted design instead of the blue 🇰🇿 flag, that is not something the shaper has any authority over. Likewise if it says that the input sequence "$1700" should be rendered one glyph at a time with the expected shapes, but that the sequence "$1701" should be substituted with a special rendering of a long ligature that may or may not look like the expected value, that is the fonts problem. The shaper is not responsible for detecting or blocking that, only for following the instructions laid out in the font.

All fonts are *software*. Not just ones that happen to have WASM instruction pages but also the ones that only use OpenType contextual substitution tables—they are all little pieces of software that can be used to do things.

**Reply by dwt** [reply]
@alerque Thanks for the clarifications, and I think I get that. It's more that I think this should be a) stated up front so others will not come to the same misconclusion and that b) I would still  hope, that the ability of the user to read what was actually written is important and that this is an important project goal.

Off the top of my hat, I can think of various ways to render fonts the way they instruct while still also following the security goal.

For example:
* Expose this information in important places like the Readme
* Very explicitly ship validated safe fonts with the engine and expose a safe mode (that users or software can invoke) to force a very reliably display. This will likely have to disable many features.
* Allow disabling all features that allow such substitutions. For example via an environment variable, probably there are better ways.

To me this seems like a case where the project should be part of a solution to this problem (that is of course not caused solely by it) and try and mitigate the problems where it can instead of just offloading the problem completely on society.

I get that displaying fonts is hard, and that the standards are written by pople who don't seem to care about about the important part they have and should play for computing to also be reliable and secure, but that seems a poor excuse to not have that as part of this project. And I know that I'm asking a lot - but as a wise person once said: "If not us, then who, if not now, then when?"

**Reply by Lewiscowles1986** [reply]
So the reason I was alerted to this at all, is that someone made a [local LLM to run via a font](https://fuglede.github.io/llama.ttf/), with this library; gives instructions to end-users about how to use (I'd presumed for non-shaping activities, maybe I'm wrong there).

https://github.com/harfbuzz/harfbuzz/discussions/4767#discussioncomment-9867834 does not help.

If someone language is so complex they need a turing complete runner; it has to be causing problems in other areas of their lives.

**Reply by khaledhosny** [reply]
> If someone language is so complex they need a turing complete runner; it has to be causing problems in other areas of their lives.

Please refrain from making thinly veiled racist comments here. If you want to understand why WASM shaper is needed, you are welcome to check its examples: https://github.com/harfbuzz/harfbuzz-wasm-examples

**Reply by Lewiscowles1986** [reply]
It's not racist to ask if something is over-complex. Recognising real life challenges, folks have; and have been relayed to me about languages I've not identified or mentioned, by individuals (when I gave no specific examples) is just believing people when they relay their struggles. Recognising that I've met many people who have highlighted challenges with their language, and sharing that wasn't intended to make anyone feel bad or question their human value.

At no point did I say anything about the people themselves. I didn't question their capacity, suggest any superiority. Racism has to come from a place where people, not tooling is insulted; so please refrain from channelling any frustration towards accusatory statements that have no basis. 

At worst it was reductive. I'll take from this that you found this inappropriate and apologize for offence caused; but please do not run around the internet, or github insulting people because they ask questions.

**Reply by khaledhosny** [reply]
You asked a question and you got an answer. Going on to make comments implying that some cultures are not worth supporting because of some perceived complexity, is naked racism.

We are not here to make a judgement call about what is what kinds of texts are worth supporting and what is not, because that is racism. We are here to find technical ways to display all texts, and find innovative ways to make this as seamless as possible.

Please focus on the technical issue at hand. No need to panic, WASM shaper is an experimental technology preview for finding more optimal ways to do things that are already possible, and it is not enabled by default anywhere.

BTW, displaying any text on a computer requires a “turing complete runner”, they are computers after all, just that some some texts are privileged that the complicity is hidden away by being built into every computer system at the lowest levels, while others have to suffer handling the complexity at a much higher level.

**Reply by Lewiscowles1986** [reply]
I Have not said cultures were not worth supporting... 

Your desire to take this into a fight is in full display. I issued you an apology, based on words and feelings you keep imagining into sentences. I've reported you to GitHub as this is unacceptable.

Nobody suggested the texts should not be supported, only that the approach of shipping a turing complete language via font is a bad idea, and likely not the best way to support that.

Did I over-extend asking what such complexity would mean? Perhaps. Perhaps I looked at it too much like an engineer.

Your follow-ups are more hateful than anything I have said, as they explicitly spew lies, accusations, and actual distasteful sentiment towards populations, exploring fictionalised hate; on the basis of disagreeing with a technology choice; while inventing things I have not said, and have explicitly refuted.

"But OpenType..." Well the record here and in examples shows that restrictions likely need to be put in-place to protect end-users of all fonts. Nothing about saying that here, says it should not also be true of OpenType. If someone would like to link a complaints thread there; I'll gladly join it. I only found out about this project because it was linked in a tech demo for embedding an LLM into a font. 

> displaying any text on a computer requires a “turing complete runner”

I've never seen evidence that turing is required for the low-bar of displaying text. I'm fairly certain it can be done (perhaps a space / compute trade-off) without. I'd re-examine history as devices to create text computationally existed long before Turing. The fact we are both, likely using turing-complete runners to display screens of text says nothing for it's necessity; nor the logical leap that every single aspect, which is within one turing machine, should be nested within others. It's not a debate line that interests me; but is indicative of missing words in statements and then imagining others. You make a wide open claim. ANY TEXT, not arbitrary text in any representation.

Not allowing arbitrary code execution over a network or any medium is the well-researched subject of many engineers and scientists. It is likely why there is a compile-time flag to enable and disable such features. There is a name for it. RCE; Remote Code Execution and it is a type of vulnerability. Which was my interest. I am ecstatic this feature is reported as off-by-default. I Remain concerned that nobody here has thought about the fact the end-user would be impacted; potentially because of a non-end-user decision. (Chrome, Firefox, etc use this)

The accepted answer gets to this just fine. This sub-thread, created after the fact, is damaging. Passion has spilled over into fanaticism; insults, poor behaviour. You keep insulting me and issuing public untrue statements about me.

Please someone lock this thread; I'm happy to remove my comments, so long as ones accusing me of racism are similarly removed.

**Reply by dwt** [reply]
@alerque Thinking about this more, maybe the alarm I'm feeling is more from the fact that it seems that the shaper is nondeterministic and or can be dependent on time. This seems to enable the attack of e.g. changing sums or terms on signed contract pdfs. (And I get that PDFs offer various other attack vectors that probably allow this).

So maybe I'm missing something, but is this actually the case? Can font rendering be non deterministic or dependent on the time? If WASM support is included, it seems highly likely that this is the case, but I haven't read the source code to understand this project.

Still, I think it important to note this particular feature of the shaper and would like to contribute by clarifying those points, so I would like some feedback if you guys will take it.

**Reply by alerque** [reply]
Once again, you're worried about this at the wrong level. If you are using a software rendering stack to view unshaped text content and relying an that for contract longevity then you are doing something else wrong. You have other problems than a possibly more flexible shaper system that isn't even enabled anywhere.

Properly generated PDFs should be already typeset. That means what you are viewing in a PDF has already been shaped and the output of the shaper *captured*. PDFs are a bizarre format with many possible ways of encoding things and yes you *can* make PDFs with unshaped content in them that would be subject to the viewers text handling software stack. If you are archiving contracts you need to make sure you are using PDFs where that is not the case and each glyph or grapheme is already resolved from the input text and it's position fixed in the output. It shouldn't depend on system fonts, and any embeded fonts should be used just for glyph shapes where at most anything the viewer does is word and character at a time, not whole lines or paragraphs. And of course you should checksum and/or sign the file. If archiving is really a big deal a rasterized version (possible to do with embeded images in the PDF) that is likewise checksummed will serve you much better.
